### PR TITLE
A simple MaxReg implementation

### PR DESCRIPTION
**Pull request by lenary**
In accordance with prophecy #55, I have written a register that stores the largest integer that you assign to it.

Questions
- [x] Should this be for all terms, or just numbers? (I think it's easier to manage expectations with just numbers)
- [x] Should it be integers or all numbers? (I coded with integers to start now, easily expanded)
- [ ] What do we do about returning `undefined` to a user? 
- [x] Do we error if someone assigns a smaller number to the register than what's in it? (I think no)
- [x] Do we error if someone assigns a non-number to the register? (I think yes)
- [x] Do we want a MinReg too? (I think maybe, but one was enough to start)

### CODE REVIEW

File: src/riak_dt_maxreg.erl [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+max_with_small_undefined(undefined, X) ->
+    X;
+max_with_small_undefined(X, undefined) ->
+    X;
+max_with_small_undefined(X, Y) ->
+    max(X, Y).
+
+%% @doc Are two `maxreg()'s structurally equal? This is not `value/1' equality.
+%% Two registers might represent the value `armchair', and not be `equal/2'. Equality here is
+%% that both registers contain the same value and timestamp.
```
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: Looks like a copypasta doc comment from LWW?

**Comment by russelldb** [comment]
In answer to you questions
1. numbers
2. integers
3. I dunno, what's wrong with zero (misleading in a usecase with negative numbers I suppose)
4. No
5. Yes
6. Yes please!

**Comment by seancribbs** [comment]
Maybe a more general "Range register"? Something that records the min, max, and latest value?

**Comment by lenary** [comment]
@rdb for 3, I wonder if we can return some kind of negative infinity. I have no idea, going to stick with undefined for the moment. As for other answers, I agree, will sort out everything. :D 

@seancribbs I'll start working on that in a second. Useful for finance stuff, where you want (max,min,earliest,latest)

**Comment by lenary** [comment]
Needs to be rebased against #71 due to tag changes.

**Comment by seancribbs** [comment]
Assigning to 2.1

**Comment by lenary** [comment]
As expected.

File: src/riak_dt_maxreg.erl [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+-ifdef(EQC).
+eqc_value_test_() ->
+    crdt_statem_eqc:run(?MODULE, 1000).
+
+%% EQC generator
+generate() ->
+    ?LET({Op, Actor}, {gen_op(), char()},
+         begin
+             {ok, Lww} = riak_dt_maxreg:update(Op, Actor, riak_dt_maxreg:new()),
```
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: lol, copy pasta naming of the register as `Lww`

File: src/riak_dt_maxreg.erl [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+-export([gen_op/0, update_expected/3, eqc_state_value/1, init_state/0, generate/0]).
+-endif.
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export_type([maxreg/0, maxreg_op/0]).
+
+-opaque maxreg() :: integer() | undefined.
```
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: We can't do this, `opaque` are treated as unique by dialyzer. This claims the type `integer() | undefined` uniquely for `maxreg()`.

File: src/riak_dt_maxreg.erl [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+-spec equal(maxreg(), maxreg()) -> boolean().
+equal(Val1, Val2) ->
+    Val1 =:= Val2.
+
+-include("riak_dt_tags.hrl").
+-define(TAG, ?DT_MAXREG_TAG).
+-define(V1_VERS, 1).
+
+% TODO: is 32 bits of integer enough? Do we require more bits?
+% TODO: do we want to be using t2b/b2t here?
```
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: I don't think we need to.

File: src/riak_dt_maxreg.erl [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+%% that both registers contain the same value.
+-spec equal(maxreg(), maxreg()) -> boolean().
+equal(Val1, Val2) ->
+    Val1 =:= Val2.
+
+-include("riak_dt_tags.hrl").
+-define(TAG, ?DT_MAXREG_TAG).
+-define(V1_VERS, 1).
+
+% TODO: is 32 bits of integer enough? Do we require more bits?
```
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: Hmmm. I don't know, if I say yes, no doubt there will be a slap down the line, so maybe use t2b and it won't matter!

Comment (lenary) [inline src/riak_dt_maxreg.erl]: now that riak_dt has helpers for this, it's pretty simple to use t2b and b2t. I'll get that done.

File: src/riak_dt_maxreg.erl (hunk 3 above)
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: Actually, I'm unsure of this now, I need to check with @reiddraper, who first brought this to my attention. Hold off for now.

File: src/riak_dt_minreg.erl [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+%% that both registers contain the same value.
+-spec equal(minreg(), minreg()) -> boolean().
+equal(Val1, Val2) ->
+    Val1 =:= Val2.
+
+-include("riak_dt_tags.hrl").
+-define(TAG, ?DT_MINREG_TAG).
+-define(V1_VERS, 1).
+
+% TODO: is 32 bits of integer enough? Do we require more bits?
```
Comment (russelldb) [inline src/riak_dt_minreg.erl]: same as for `maxreg`, I guess you know that now.

File: src/riak_dt_minreg.erl [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+-ifdef(EQC).
+eqc_value_test_() ->
+    crdt_statem_eqc:run(?MODULE, 1000).
+
+%% EQC generator
+generate() ->
+    ?LET({Op, Actor}, {gen_op(), char()},
+         begin
+             {ok, Lww} = riak_dt_minreg:update(Op, Actor, riak_dt_minreg:new()),
```
Comment (russelldb) [inline src/riak_dt_minreg.erl]: as for `maxreg` I know it is a small thing, but it is confusing on first read.

File: src/riak_dt_rangereg.erl [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc
```
Comment (russelldb) [inline src/riak_dt_rangereg.erl]: All copy pasta, please add a real `@doc` comment

File: src/riak_dt_rangereg.erl [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+fww(undefined, Other) ->
+  Other;
+fww(Other, undefined) ->
+  Other;
+fww({VA,TsA}, {_VB,TsB}) when TsA < TsB ->
+  {VA,TsA};
+fww({_VA,TsA}, {VB,TsB}) when TsA > TsB ->
+  {VB,TsB};
+fww(A, B) ->
+  min(A,B). %% some kind of tiebreaker.
```
Comment (russelldb) [inline src/riak_dt_rangereg.erl]: Please document this arbitrator in either `merge/2` or the module edoc.

File: src/riak_dt_rangereg.erl [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+%% that both registers contain the same value.
+-spec equal(rangereg(), rangereg()) -> boolean().
+equal(Val1, Val2) ->
+    Val1 =:= Val2.
+
+-include("riak_dt_tags.hrl").
+-define(TAG, ?DT_RANGEREG_TAG).
+-define(V1_VERS, 1).
+
+% TODO: compression?
```
Comment (russelldb) [inline src/riak_dt_rangereg.erl]: Please measure briefly, I doubt there is any value to compressing vs. cpu time, but why not check?

File: src/riak_dt_rangereg.erl [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+-ifdef(EQC).
+eqc_value_test_() ->
+    crdt_statem_eqc:run(?MODULE, 1000).
+
+%% EQC generator
+generate() ->
+    ?LET({Op, Actor}, {gen_op(), char()},
+         begin
+             {ok, Lww} = riak_dt_rangereg:update(Op, Actor, riak_dt_rangereg:new()),
```
Comment (russelldb) [inline src/riak_dt_rangereg.erl]: What's in a name?

File: src/riak_dt_rangereg.erl (hunk 8 above)
Comment (lenary) [inline src/riak_dt_rangereg.erl]: Done

File: src/riak_dt_rangereg.erl (hunk 9 above)
Comment (lenary) [inline src/riak_dt_rangereg.erl]: Done both

**Comment by lenary** [comment]
TODO:
- [x] stats callbacks (required by kv)
- [x] error if non-integral value is assigned to the register.

**Comment by lenary** [comment]
Ok, I think the last things to do are to benchmark the compression (vs without), and to look at some interface details.

**Comment by lenary** [comment]
Benchmarking compression (via `riak_dt:to_binary`) vs nowt (via `term_to_binary`):

```
         compd        uncompd
maxreg (and minreg):
time:    4.04         0.39
size:    [5..8]       [5..8]

rangereg:
time:    9.49         0.90
size:    [32..43]     [32..44]
```

so i guess no compression then

File: src/riak_dt_maxreg.erl (hunk 3 above)
Comment (seancribbs) [inline src/riak_dt_maxreg.erl]: @russelldb You are correct. Opaque types must be composite types (records usually). They can't be a union type of two or more fundamental types.

File: src/riak_dt_maxreg.erl [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+-spec value(term(), maxreg()) -> term().
+value(_, V) ->
+    value(V).
+
+%% @doc Assign a `Value' to the `maxreg()'
+-spec update(maxreg_op(), riak_dt:actor(), maxreg()) ->
+                    {ok, maxreg()} | nonintegral_error().
+update({assign, Value}, _Actor, OldVal) when is_integer(Value) ->
+    {ok, merge(OldVal, Value)};
+update({assign, Value}, _Actor, _OldVal) ->
```
Comment (russelldb) [inline src/riak_dt_maxreg.erl]: I wonder about this, what do you think @seancribbs? For me, erlang would generally not have this case, and just let the caller crash.

**Comment by russelldb** [comment]
Most CRDTs also throw in the to/from binary roundtrip test (see https://github.com/basho/riak_dt/blob/develop/src/riak_dt_orswot.erl#L562), would you mind, awfully?

File: src/riak_dt_maxreg.erl (hunk 12 above)
Comment (seancribbs) [inline src/riak_dt_maxreg.erl]: You could let the caller crash, yes, or use `error(badarg)`.

**Comment by russelldb** [comment]
I'm plus one on this, but I'm not going to plus one until we branch for 2.0rc.

Good job, man! Welcome back!

**Pull request closed without merging**